frontend/ui/widget/readerprogress.lua

```diff
@@ -0,0 +1,481 @@
+            padding_span,
+            line_container,
+            padding_span,
+            line_container,
+            padding_span,
+            header_title_right,
+            padding_span,
+            line_container,
+            padding_span,
+        },
+        VerticalSpan:new{ width = Screen:scaleBySize(5), height = height },
+    }
+end
+
+function ReaderProgress:genWeekStats()
+    local all_day_period = 86400
+    local date_format
+    local date_format_show
+    local select_day_time
+    local now_time = os.time()
+    local screen_width = Screen:getWidth()
+    local screen_height = Screen:getHeight()
+    local height = Screen:scaleBySize(60)
+    local STATS_DAY = 7
+    local statistics_container = CenterContainer:new{
+        dimen = Geom:new{ w = screen_width , h = height },
+    }
+    local statistics_group = VerticalGroup:new{ align = "left" }
+    local max_week_time = -1
+    for _, v in pairs(progress_dates) do
+        if v.read > max_week_time then max_week_time = v.read end
+    end
+
+    local top_padding_span = HorizontalSpan:new{ width = Screen:scaleBySize(15) }
+    local top_span_group = HorizontalGroup:new{
+        align = "center",
+        LeftContainer:new{
+            dimen = Geom:new{ h = Screen:scaleBySize(30) },
+            top_padding_span
+        },
+    }
+    table.insert(statistics_group, top_span_group)
+
+    for i = 1, STATS_DAY , 1 do
+        date_format = os.date("%Y-%m-%d (%a)" , now_time - all_day_period * (i -1))
+        if progress_dates[date_format] ~= nil then
+            select_day_time = progress_dates[date_format].read
+        else
+            select_day_time = 0
+        end
+        date_format_show = os.date("%A (%d.%m)" , now_time - all_day_period * (i - 1))
+        local total_group = HorizontalGroup:new{
+            align = "center",
+            padding = 2,
+            LeftContainer:new{
+                dimen = Geom:new{ w = screen_width , h = height / 3 },
+                TextWidget:new{
+                    padding = 2,
+                    text = date_format_show .. " - " .. util.secondsToClock(select_day_time, true),
+                    face = Font:getFace("ffont", 16),
+                },
+            },
+        }
+        local titles_group = HorizontalGroup:new{
+            align = "center",
+            LeftContainer:new{
+                dimen = Geom:new{ w = screen_width , h = height / 3 },
+                ProgressWidget:new{
+                    width = (screen_width * 0.005) + (screen_width * 0.9 * select_day_time / max_week_time),
+                    height = Screen:scaleBySize(14),
+                    percentage = 1.0,
+                    ticks = nil,
+                    last = nil,
+                    margin_h = 0,
+                    margin_v = 0,
+                }
+            },
+        }
+        local padding_span = HorizontalSpan:new{ width = Screen:scaleBySize(15) }
+        local span_group = HorizontalGroup:new{
```

Comment:
`padding_span` and `span_group` are static widgets that doesn't change with the loop. So it's better to create them outside of the loop and reuse the same one in the loop.